source/Octopus.Tentacle.Tests.Integration/Support/TentacleConfigurationsAttribute.cs

```diff
@@ -0,0 +1,121 @@
+                    TentacleVersions.v5_0_12_AutofacServiceFactoryIsInShared,
+                    TentacleVersions.v6_3_417_LastWithScriptServiceV1Only, // the autofac service is in tentacle, but tentacle does not have the capabilities service.
+                    TentacleVersions.v7_0_1_ScriptServiceV2Added
+                });
+            }
+
+            if (testScriptIsolationLevel || testScriptsInParallel)
+            {
+                versions.AddRange(new[]
+                {
+                    TentacleVersions.v6_3_417_LastWithScriptServiceV1Only,
+                    TentacleVersions.v7_0_1_ScriptServiceV2Added // Testing against v1 and v2 script services
+                });
+            }
+
+            List<bool?> stopPortForwarderAfterFirstCallValues =
+                testStopPortForwarderAfterFirstCall
+                    ? new List<bool?> {true, false}
+                    : new List<bool?> {null};
+
+            List<RpcCall?> rpcCalls = testRpcCalls
+                ? new List<RpcCall?> {RpcCall.FirstCall, RpcCall.RetryingCall}
+                : new List<RpcCall?> {null};
+
+            List<RpcCallStage?> rpcCallStages = testRpcCallStages
+                ? new List<RpcCallStage?> {RpcCallStage.Connecting, RpcCallStage.InFlight}
+                : new List<RpcCallStage?> {null};
+
+            List<ScriptIsolationLevel?> scriptIsolationLevels = testScriptIsolationLevel
+                ? new List<ScriptIsolationLevel?> {ScriptIsolationLevel.FullIsolation, ScriptIsolationLevel.NoIsolation}
+                : new List<ScriptIsolationLevel?> {null};
+
+            List<ScriptsInParallelTestCase?> scriptsInParallelTestCases = testScriptsInParallel
+                ? new List<ScriptsInParallelTestCase?>
+                {
+                    // Scripts with different mutex names can run at the same time.
+                    ScriptsInParallelTestCase.FullIsolationDifferentMutex,
+                    // Scripts with the same mutex name can run at the same time if they both has no isolation.
+                    ScriptsInParallelTestCase.NoIsolationSameMutex
+                }
+                : new List<ScriptsInParallelTestCase?> {null};
+
+            return (from tentacleType in tentacleTypes
+                    from halibutType in halibutTypes
+                    from version in versions.Distinct()
+                    from stopPortForwarderAfterFirstCall in stopPortForwarderAfterFirstCallValues
+                    from rpcCall in rpcCalls
+                    from rpcCallStage in rpcCallStages
+                    from scriptIsolationLevel in scriptIsolationLevels
+                    from scriptsInParallelTestCase in scriptsInParallelTestCases
+                    select new TentacleConfigurationTestCase(
+                        tentacleType,
+                        halibutType,
+                        version,
+                        stopPortForwarderAfterFirstCall,
```

Comment:
I don't think most of these bools belong on here.

We had the same problem, of needing additional params, in halibut. This was solved in `LatestClientAndLatestServiceTestCasesAttribute` using `additionalParameters`.

An alternative is to use `ValuesOfType` (which is defined in halibut). You give a type reference to an `IEnumerable`. ValuesOfType doesn't yet have any support for passing parameters. Although a copy of it in tentacle could.

Reply:
Ah cool. didn't notice the solution in `LatestClientAndLatestService...` attribute, that's most of what I wanted. I don't like that you still have to declare multiple copies of the attribute to get all permutations of an enum etc, but coupled with something like the `ValuesOfType` could make it nicer.

Rather than push back this PR to do the rework, I reckon I put it on my sharpening list and do this when I tackle your other suggestion (i.e. `ClientAndTentacleBuilder` takes the whole `TentacleConfigurationTestCase` object)

Reply:
After some more discussion with @LukeButters, I have refactored the test case attribute class to remove most of bools and config values that were specific to very small sets of tests.

There is now an `additionalParameterTypes` field, which takes an array of `Type` objects. The attribute assumes these types are either enums or implement `IEnumerable`, then generates the necessary test cases to cover all permutations.